### PR TITLE
As operator

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Implements `as` operator as suggested in #296.

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -669,6 +670,12 @@ module ts {
         right: Expression;
     }
 
+    export interface AsExpression extends Expression {
+        left: Expression;
+        asToken: Node;
```
Comment (mhegazy) [inline src/compiler/types.ts]: why do you need this token?

Comment (ahejlsberg) [inline src/compiler/types.ts]: Don't think we need to store the `as` token.

File: tests/baselines/reference/asOperator1.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+var as = 43;
+var x = undefined as number;
+var y = (null as string).length;
+var z = Date as any as string;
+
+
+//// [asOperator1.js]
+var as = 43;
+var x = undefined;
+var y = (null).length;
```
Comment (CyrusNajmabadi) [inline tests/baselines/reference/asOperator1.js]: We should do the same thing we do with parenthesized type assertions, and not emit the parens when we can.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -7264,6 +7264,16 @@ module ts {
...
+            return checkAssertion(node, exprType, targetType);
+        }
+
+        function checkAsExpression(node: AsExpression) {
+            let exprType = checkExpression(node.left);
+            let targetType = getTypeFromTypeNode(node.right);
+            return checkAssertion(node, exprType, targetType);
+        }
+
+        function checkAssertion(node: Node, exprType: Type, targetType: Type) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: You might consider having both `SyntaxKind.TypeAssertionExpression` and `SyntaxKind.AsExpression` dispatch directly to this method with a `node` parameter of type `TypeAssertionExpression | AsExpression`.

File: tests/baselines/reference/asOperator1.js (hunk 2 above)
Comment (ahejlsberg) [inline tests/baselines/reference/asOperator1.js]: Indeed. This is handled by the `emitParenExpression` in emitter.ts.

**Comment by DanielRosenwasser** [comment]
Can we have a few tests for templates in `tests/cases/.../templates/`?

```TypeScript
var x = `${123 + 456 as number}`;
var y = `leading ${123 + 456 as number}`;
var y = `${123 + 456 as number} trailing`;
```

```TypeScript;
var x = `Hello ${123} World` as string;
var y = `Hello` as string;
var z = 1 + `${1} end of string` as string;
```

```TypeScript
declare function tag(...x: any[]): any;

var x = tag `Hello ${123} World` as string;
var y = tag `Hello` as string;
```

**Comment by yuit** [comment]
Should we add formating rule for ```as```? 
For example:
```
var x = 42   as string;
```
will become
```
var x = 42 as string;
```

**Comment by yuit** [comment]
Could we add some tests:
```
var a = 20;
var b = a as string;
var as = "hello";
var as1 = as as string;
```

**Comment by tinganho** [comment]
I'm just wondering if we can support inferred type assertions and if it is a good idea?

Instead of:
```
(foo as Foo).(bar as Bar).text
```
We could just use:
```
foo.bar.text
```

And it will infer the type by looking at the last property.

**Comment by DanielRosenwasser** [comment]
> Should we add formating rule for `as`? 

If by formatting rule, you mean a restriction, yes (good catch!), but not related to the example you just gave (nowhere else in the language do we differentiate whitespace on the same line). Consider the following:

```TypeScript
class Foo { }
declare function as(...args: any[]);

// Example 1
var x = 10
as `Hello world`

// Example 2
var y = 20
as(Foo);
```

Example 1 is not as much of a problem; you can't use a template string as a type.

Example 2 suffers from potentially the same problem as #2995. `as` would ordinarily be a function call with a constructor function, but here it is a type assertion on `20` to `Foo`.

**Comment by DanielRosenwasser** [comment]
@yuit Ah, by formatting rule, you meant in the LS - still, glad we caught this.

**Comment by yuit** [comment]
@DanielRosenwasser  yes, I mean from LS side similar to:

```
function       foo() {}
```
become
```
function foo() {}
```

Though you example will be a good one to add as well.

File: src/compiler/parser.ts [hunk 4]
```diff
@@ -2818,7 +2821,12 @@ module ts {
                     break;
                 }
 
-                leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                if (token === SyntaxKind.AsKeyword) {
+                    parseTokenNode();
+                    leftOperand = makeAsExpression(leftOperand, parseType());
+                } else {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: `else` on the next line.

**Comment by RyanCavanaugh** [comment]
Any other feedback?

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -2818,7 +2821,16 @@ module ts {
                     break;
                 }
 
-                leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                if (token === SyntaxKind.AsKeyword) {
+                    if (canParseSemicolon()) {
```
Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: just add a comment for this case.

File: src/compiler/parser.ts [hunk 6]
```diff
@@ -2818,7 +2821,16 @@ module ts {
                     break;
                 }
 
-                leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                if (token === SyntaxKind.AsKeyword) {
+                    if (canParseSemicolon()) {
+                        break;
+                    } else {
+                        parseTokenNode();
```
Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: just call nextToken()   since you're not using the result of htis.

File: src/compiler/types.ts [hunk 7]
```diff
@@ -669,6 +672,12 @@ module ts {
         right: Expression;
     }
 
+    export interface AsExpression extends Expression {
+        expression: Expression;
+        asToken: Node;
```
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: remove.

File: src/compiler/parser.ts [hunk 8]
```diff
@@ -2818,7 +2821,20 @@ module ts {
...
 
-                leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                if (token === SyntaxKind.AsKeyword) {
+                    // Make sure we *do* perform ASI for constructs like this:
+                    // var x = foo
+                    // as (Bar)
+                    // This should be parsed as an initialized variable, followed by a function call to 'as' with the argument 'Bar'
+                    if (canParseSemicolon()) {
+                        break;
+                    } else {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: `else` on the next line

File: src/compiler/parser.ts [hunk 9]
```diff
@@ -2818,7 +2821,20 @@ module ts {
...
+                        nextToken();
+                        leftOperand = makeAsExpression(leftOperand, parseType());
+                    }
+                } else {
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: `else` on the next line

File: src/compiler/parser.ts [hunk 10]
```diff
@@ -2818,7 +2821,20 @@ module ts {
                     break;
                 }
 
-                leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                if (token === SyntaxKind.AsKeyword) {
+                    // Make sure we *do* perform ASI for constructs like this:
+                    // var x = foo
+                    // as (Bar)
+                    // This should be parsed as an initialized variable, followed by a function call to 'as' with the argument 'Bar'
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: with the parenthesized type 'Bar'

Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Also consider wrapping this

**Comment by DanielRosenwasser** [comment]
You're currently not contextually typing the left-hand side of the `as` operator.

```TypeScript
var x = (v => v) as (x: number) => number
```

So `v` gets typed as `any` right now. This means that

```TypeScript
var x = (v => v) as (x: number) => string
```

currently typechecks without a problem even though

```TypeScript
var x = <(x: number) => string>(v => v)
```

gives an error.

**Comment by DanielRosenwasser** [comment]
Can we also have the following tests to demonstrate left-associativity of `as`?

**<sub>`asOperatorAssociativity01.ts`</sub>**
```TypeScript
var x = 10 as number as any as string // should be okay
```

**<sub>`asOperatorAssociativity02.ts`</sub>**
```TypeScript
var y = 10 as string as number; // should error
```

**Comment by RyanCavanaugh** [comment]
Anything else?

**Comment by DanielRosenwasser** [comment]
Are we leaving the services layer to a second pass? I can think of at least:

* Formatting tests
* Does `as` get occurrence highlighting to point out type assertion locations? (kind of useful; it's the point of `static_cast` etc. in C++).
* Completion list tests for arrow functions:
```TypeScript
class C<T> {
    constructor() {
        // C and T should show up below.
        let f = x => x as /**/ 
```

```TypeScript
type A = 
namespace n {
    type B = any;
    // A, B, T, and n should show up below.
    let f = <T>(x: any) => x as /**/ 
```

**Comment by DanielRosenwasser** [comment]
I am working on a change such that `interface` must be followed by an identifier on the same line to be considered the start of an interface declaration.

But `as` is only a type assertion if it follows an identifier on the same line. So there's an ambiguity.

It seems like the appropriate thing to do is to treat it as an interface declaration named `as`.

You'll have to account for this with some tests like:

```TypeScript
interface as { }
```

```TypeScript
interface
as
{ }
```

```TypeScript
interface
as({ })
```

File: src/compiler/parser.ts [hunk 11]
```diff
@@ -2818,7 +2821,23 @@ module ts {
...
                 }
 
-                leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                if (token === SyntaxKind.AsKeyword) {
+                    // Make sure we *do* perform ASI for constructs like this:
+                    //    var x = foo
+                    //    as (Bar)
+                    // This should be parsed as an initialized variable, followed
+                    // by a function call to 'as' with the argument 'Bar'
+                    if (canParseSemicolon()) {
```
Comment (JsonFreeman) [inline src/compiler/parser.ts]: Can you just check scanner.hasPrecedingLineBreak instead? I don't really like canParseSemicolon that much.

File: src/compiler/parser.ts [hunk 12]
```diff
@@ -2818,7 +2821,23 @@ module ts {
...
+                        break;
+                    }
+                    else {
+                        nextToken();
+                        leftOperand = makeAsExpression(leftOperand, parseType());
```
Comment (JsonFreeman) [inline src/compiler/parser.ts]: There is an ambiguity with union types worth calling out here:
```ts
x as Foo | Bar
```
This is intended to be a type assertion to a union type, and not a bitwise OR where the first operand is a type assertion. Seems like you do the right thing here, but just wanted to bring it up.

File: src/compiler/emitter.ts [hunk 13]
```diff
@@ -1790,8 +1790,8 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
             }
 
             function skipParentheses(node: Expression): Expression {
-                while (node.kind === SyntaxKind.ParenthesizedExpression || node.kind === SyntaxKind.TypeAssertionExpression) {
-                    node = (<ParenthesizedExpression | TypeAssertion>node).expression;
+                while (node.kind === SyntaxKind.ParenthesizedExpression || node.kind === SyntaxKind.TypeAssertionExpression || node.kind === SyntaxKind.AsExpression) {
+                    node = (<ParenthesizedExpression | AssertionExpression>node).expression;
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Need to update parenthesizeForAccess as well.

**Comment by mhegazy** [comment]
@RyanCavanaugh is this ready to go in?

**Comment by RyanCavanaugh** [comment]
I'm merging this up with the JSX work, which should have a PR in a day or so

**Comment by mhegazy** [comment]
closing in favor of https://github.com/Microsoft/TypeScript/pull/3564

**Pull request closed without merging**